projects/angular/src/utils/list-key-navigation/key-navigation-list.ts

```diff
@@ -0,0 +1,144 @@
+  initializeKeyListItems,
+  KeyNavigationCode,
+  setActiveKeyListItem,
+  validKeyNavigationCode,
+} from '@cds/core/internal';
+
+import { getNextKeyListItem } from './utils';
+
+export interface KeyNavigationListConfig {
+  keyListItems: string;
+  layout: 'both' | 'horizontal' | 'vertical';
+  manageFocus: boolean;
+  manageTabindex: boolean;
+  loop: boolean;
+  dir: string | null;
+}
+
+@Directive({
+  selector: '[clrKeyNavigationList]',
+})
+export class ClrKeyNavigationList {
+  private get listItems(): NodeListOf<HTMLElement> {
+    return this.host.nativeElement.querySelectorAll(
+      [
+        'a[href]',
+        'area[href]',
+        'input:not([disabled])',
+        'button:not([disabled])',
+        'select:not([disabled])',
+        'textarea:not([disabled])',
+        'iframe',
+        'object',
+        'embed',
+        '*[tabindex]', // -1 tabindex is a focusable element and needed for keyboard navigation
+        '*[contenteditable=true]',
+        '[role=button]:not([disabled])',
+      ].join(',')
+    );
+  }
+
+  private config: KeyNavigationListConfig;
+
+  constructor(
+    private host: ElementRef<HTMLElement>,
+    private renderer: Renderer2,
+    @Inject('ClrListKeyNavigation') config: Partial<KeyNavigationListConfig>
+  ) {
+    this.config = {
+      keyListItems: 'keyListItems',
+      layout: 'vertical',
+      manageFocus: true,
+      manageTabindex: true,
+      loop: false,
+      dir: this.host.nativeElement.getAttribute('rtl'),
+      ...config,
+    };
+  }
+
+  @Input()
+  set configuration(config: Partial<KeyNavigationListConfig>) {
+    this.config = {
+      ...this.config,
+      ...config,
+    };
+  }
+
+  ngAfterViewInit() {
+    this.initializeTabIndex();
+    this.renderer.listen(this.host.nativeElement, 'click', this.clickItem.bind(this));
+    this.renderer.listen(this.host.nativeElement, 'keydown', this.focusItem.bind(this));
+  }
+
+  private initializeTabIndex() {
+    if (this.config.manageFocus && this.config.manageTabindex) {
+      initializeKeyListItems(this.listItems);
+    }
+  }
+
+  private clickItem(e: Event) {
+    const activeItem = this.getActiveItem(e);
+    if (activeItem) {
+      this.setActiveItem(e, activeItem);
+    }
+  }
+
+  private focusItem(e: KeyboardEvent) {
+    if (validKeyNavigationCode(e)) {
+      const activeItem = this.getActiveItem(e);
+      if (activeItem) {
+        const { next, previous } = getNextKeyListItem(activeItem, Array.from(this.listItems), {
+          ...this.config,
+          code: e.code as KeyNavigationCode,
+        });
+
+        if (next !== previous) {
+          // todo: test for only emiting when position has changed
+          this.setActiveItem(e, this.listItems[next], this.listItems[previous]);
+        }
+      }
+    }
+  }
+
+  private getActiveItem(e: Event) {
+    return Array.from(this.listItems).find(
+      c => c === (e.target as HTMLElement).closest(this.listItems[0].tagName.toLocaleLowerCase()) ?? c === e.target
```

Comment:
See my comment about `listItems` being changed to a method. This code is unnecessarily expensive because it runs the DOM query several times. Please look at all uses of the `listItems` property for easy optimizations.